### DISCUSSION TITLE
How to use static and dynamic libraries at the same time?

### DISCUSSION THREAD

**Discussion by YuHuanTin** [microsoft/vcpkg#33256] [Q&A]
I am very confused as I can't seem to find a way to add both static and dynamic libraries to my cmake project?

- I've tried using the `set(VCPKG_TARGET_TRIPLET "x86-windows-static")` command before I `find_package` the libraries I want to use, but this command seems to be sometimes useful and sometimes useless (so I'm not sure if it's the only or the correct way for me to specify static or dynamic libraries) way).
    In my tests, I have both the `x86-windows` and `x86-windows-static` versions of `openssl` and `boost-locale`, and when I use the
    ```
    find_package(xxx)
    ...
    set(VCPKG_TARGET_TRIPLET "x86-windows-static")
    find_package(OpenSSL REQUIRED)
    ```
    But when I want to use the `x86-windows-static` version of `openssl`, everything works fine and my `cmake` correctly recognizes the `x86-windows-static` version of `openssl` 
    ![image](https://github.com/microsoft/vcpkg/assets/51024916/40eda620-1afd-4675-b08c-183bc9292163)

    But when trying to use the `static` version of `boost-locale`, everything sucks because it recognizes the dynamic library version instead of the static version I want!
    ```
    find_package(xxx)
    ...
    set(VCPKG_TARGET_TRIPLET "x86-windows-static")
    find_package(Boost REQUIRED COMPONENTS locale)
    ```  
    ![image](https://github.com/microsoft/vcpkg/assets/51024916/ba300173-a232-4ba0-896a-4eef1aa66f09)

So I'm asking if there is a way to add both dynamic libraries (not shown in the test because I only want to statically link `boost-locale`) and static libraries in `cmake`?

**Comment by Neumann-A** [comment]
> I am very confused as I can't seem to find a way to add both static and dynamic libraries to my cmake project?

You cannot mix triplets! You cannot select triplets by setting `VCPKG_TARGET_TRIPLET` before `find_package()`! Everything gets decided when `project()` is called.

To use a custom combination of static/dynamic libraries you need to write your own triplet (provided via [overlay triplets](https://learn.microsoft.com/en-us/vcpkg/users/examples/overlay-triplets-linux-dynamic)) and apply per [port customization](https://learn.microsoft.com/en-us/vcpkg/users/triplets#per-port-customization)

**Reply by YuHuanTin** [reply]
Thanks for your reply, I have read this article but I don't know how to use that cmake file `x64-linux-dynamic.cmake`? Can you give me a small example?

Taking the above `boost-locale` component as an example, I created the `x64-linux-dynamic.cmake` file and then filled in the following instructions

```
set(VCPKG_TARGET_ARCHITECTURE x64)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE static) # This changed from static to dynamic

set(VCPKG_CMAKE_SYSTEM_NAME Linux)
```

But I don't know what to do after the second step `Step 2: Use --overlay-triplets to build dynamic libraries`.

And if I use the command `vcpkg install boost-locale:x86-windows-static` should I change `VCPKG_CRT_LINKAGE` to `static` in the above code?

I've found that the `vcpkg\triplets\community\x86-windows-static.cmake` file seems to be what the article is showing, but does that have anything to do with the fact that I want to use a static `boost-locale`?

**Reply by YuHuanTin** [reply]
I seem to have figured out how to do it
- I need to create a folder with any name (`triplets-custom` for example) and put in `x86-windows-boost-locale-static.cmake`
    with the following contents
    ```
    set(VCPKG_TARGET_ARCHITECTURE x86)
    set(VCPKG_CRT_LINKAGE dynamic)
    set(VCPKG_LIBRARY_LINKAGE dynamic)

    if (${PORT} MATCHES "boost-locale")
        set(VCPKG_LIBRARY_LINKAGE static)
    endif ()
    ```

- Then you need to use
    ``vcpkg install boost-locale:x86-windows-boost-locale-static --overlay-triplets=triplets-custom``
    `vcpkg install openssl:x86-windows-boost-locale-static --overlay-triplets=triplets-custom` 
    command builds the libraries (where `openssl` is a dynamic library and `boost-locale` is built as a static library).

- Finally we write in `cmakelist.txt` to include only the static version of `boost-locale` and the rest are dynamic libraries
    ```
    set(VCPKG_OVERLAY_TRIPLETS "triplets-custom")
    set(VCPKG_TARGET_TRIPLET "x86-windows-boost-locale-static")
    project(ProjectName)

    set(CMAKE_CXX_STANDARD 20)
    find_package(OpenSSL REQUIRED)
    find_package(Boost REQUIRED COMPONENTS locale)

    link_libraries(OpenSSL::SSL OpenSSL::Crypto)
    link_libraries(Boost::boost Boost::locale)
    ```


I find this very cumbersome, is there a better way?

**Reply by Neumann-A** [reply]
> I find this very cumbersome, is there a better way?

Use a manifest (vcpkg.json) with the dependencies listed and the overlay specified. Set `VCPKG_TARGET_TRIPLET` via cmake presets? 

Apart from that. No.

**Reply by YuHuanTin** [reply]
ok, thanks for the reply, I will go and try the vcpkg.json

**Comment by davidhunter22** [comment]
I have a variant on this issue. I have a large project but it boils down to the following.
I am using project here to refer to a Visual Studio solution containing two VS projects all generated by CMake and using vcpkg.

In one project I want to use the static version of a vckpg package in my case pugixml so `target_link_libraries(foo PUBLIC pugixml::static)`
in another project I want to use a dynamic version of the same package so `target_link_libraries(foo PUBLIC pugixml::pugixml)`

Note both projects are completely independent and we have good reasons to want to do this.
My understanding is that vcpkg can't support this in any way. I have seen comments elsewhere that the root of the issue is that vcpkg bakes in the "static" versus "dynamic" of a library into the triplet. I can understand why you want the platform, architecture, compiler, ... to be consistent in a build but I see zero reason why you should enforce static/dynamic at the same time. As my case shows it is absolutely reasonable to use both the static and dynamic version of a library in the same build tree.

At present we don't use vcpkg for situations like this because of the design decisions that seem to make this impossible, which is a great shame.

**Reply by YuHuanTin** [reply]
Do you mean that in the same cmake project using two
`target_link_libraries(foo PUBLIC pugixml::pugixml)`
`target_link_libraries(foo PUBLIC pugixml::static)`

I can't seem to do this in vcpkg?

Of course I can assume that you are using the compiler options to control whether to use the static or dynamic version, and then use vcpkg to import both the static and dynamic versions of the library <- is that what you expect?

**Reply by davidhunter22** [reply]
No, that would be trying to use a static and a dynamic version of the same library in a single target binary which I would assume would failed in some way. I want to do 
```
target_link_libraries(foo_1 PUBLIC pugixml::pugixml)
target_link_libraries(foo_2 PUBLIC pugixml::static)
```
The problem is that when using a manifest I can either use the `x64-windows` triplet which will build `pugixml::pugixml` library or I can use the `x64-windows-static` triplet that will build the `pugixml:static` library. But I can't use both triplets and have both available

**Reply by YuHuanTin** [reply]
Yes .... There is no way to use both ? (I'm not really sure.)

**Comment by RelicOfTesla** [comment]
How can I do static and dynamic same time in vcpkg.json??

**Reply by YuHuanTin** [reply]
hi, you can try this way (specify platform 'static')
![Untitled](https://github.com/user-attachments/assets/d5b918cf-4b1f-4931-9e0e-c1ee18d17c80)

**Reply by RelicOfTesla** [reply]
I have tried, the keyword "platform" only a build condition , not a build parameter. like "supports"..

**Reply by dg0yt** [reply]
Linkage is controlled by the triplets. To build a port with different linkage, you must use a custom triplet file.
https://learn.microsoft.com/en-us/vcpkg/users/triplets#per-port-customization

**Reply by YuHuanTin** [reply]
Gosh, I seem to have misunderstood the meaning of the 'platform' field, it doesn't build according to static but filters when selecting a 'triplet'!

**Comment by etheory** [comment]
This is possibly the biggest design flaw with vcpkg ever. Not being able to do this (mix dynamic and static libraries provided by it), simply, without the need for the huge pain of custom triplets, make the entire thing feel poorly designed. That, along with no ability to easily specify which version of things you want to use. Is this hard set, or are these shortcomings going to be resolved at some point in the future? I REALLY want to like vcpkg, but honestly, it still seems to be quite a mess still. It only works for me in the simplest of cases. As soon as I want to use it on a reasonably sized project, I have to go elsewhere.

**Comment by hansingt** [comment]
I have a (somewhat) different case, which I think, I cannot resolve at all because of this limitation:

I have a custom port for a dependency not managed by vcpkg. This port can provide dynamic libraries only. Thus, I've added `!(static & staticcrt)` in the `supports` field of the port. That's all fine.

The problem now is, that this port has a dependency to OpenSSL, **which it requires to be provided as dynamic library**. As far as I've understood, the triplets only work on "project" level, not on port level, right? 

Thus, the **users** of that port need to know this limitation and have to create a custom triplet, which ensures that OpenSSL is always built dynamic, right? This is bad, especially, because this prevents them from statically linking OpenSSL for the rest of the project (which works perfectly well)..

Have I missed something? Is it possible to somehow declare the dependency dynamic library to be dynamic as well without affecting the overall project?

**Reply by dg0yt** [reply]
Do I understand correctly that you want to simultaneously use static and dynamic libraries of the same port (openssl)?

**Reply by hansingt** [reply]
In fact I don't even know, whether the **users** of my port will want to use a dynamic and static version of OpenSSL at the same time or whether they want to (directly) use OpenSSL at all.

I'm writing a port for a library which requires OpenSSL to be linked dynamically, but I cannot express this using vcpkg, or at least I'm not aware of any mechanism to express this in the port. 

I can only express, that my port has a dependendency to OpenSSL, but not whether this dependency is dynamic or static.. This is decided only by the triplet of the project **using** my port.

**Reply by dg0yt** [reply]
> I can only express, that my port has a dependendency to OpenSSL, but not whether this dependency is dynamic or static.. This is decided only by the triplet of the project using my port.

AFAICT it is really by design: The triplet (file) decides about the linkage (e.g. of openssl), not the reverse dependencies (your port). In general, ports shall not restrict users.

What your port can do is check which linkage of openssl it finds at build time, and fail with a actionable message if it is the variant which the package doesn't want to support.
